### ISSUE TITLE
C.120 Understanding "Do not use inheritance when simply having a data member will do."

### ISSUE THREAD

**Issue by CompCodeNet** [isocpp/CppCoreGuidelines#1963]
1. If inheritance is avoided in favor of an object as a data member (as applicable), then in order to expose functions of the data member object for simplicity, must wrapper functions be used? 
2. If No. 1 is true, that is considerable additional code that could objectively be considered superfluous - agree?

**Comment by jwakely** [comment]
It's not superfluous if it avoids unwanted coupling and other unwanted effects like derived-to-base conversions, ADL, slicing etc.

**Comment by CompCodeNet** [comment]
Thank you, Jonathan for confirming that this is reasonable. At the same time, I propose, that some might choose to ignore this guideline in some cases. and wait until the language has support for automatically exposing functions of object data members at the containing object interface (for example `expose public SomeClass _someMember;`.

**Comment by jwakely** [comment]
The guidelines cannot cover all situations. If waiting indefinitely for a hypothetical language extension that isn't on the agenda is the right call for your project, that's your call.

The guideline seems correct as written for today's C++ language.

**Comment by CompCodeNet** [comment]
Request editor's call.

**Comment by CompCodeNet** [comment]
This also relates to whether wrapper functions to expose data object members would ever be needed because if the data member is the correct decision based on overall application design, then exposing the functions might not make sense. In that case, my syntax suggestion would also not be applicable because the usefulness of wrappers would not apply.

I think that the key with C.120 is not whether "a data member will do" 🙂. As we know, inheritance is (very 🙂) useful but not essential, so one could always design around inheritance and use data members - or vice-verse one can force inheritance by deciding to make constructor or other functions protected etc. The real decision is whether inheritance efficiently and intelligently serves the goals of the application. This relates more to the type as a concept, and the concept itself, again, relates to the goals of the application. The language is here to serve us, we are not here to serve the language 🙂

**Comment by CompCodeNet** [comment]
An example is helpful. However, the starting point should be explaining the purpose of inheritance as a language capability. As discussed in previous post, the purpose of inheritance is to abstract or extend types that fulfill a conceptual role. So, one would only want to inherit (including multiple) if the resulting (derived) exposed interface fits the concept of the type. I.e. What you are inheriting is in line with the type's concept. From this understanding, one can then distinguish between internal members that support the implementation of the class's concept vs. exposed functions that manifest the class's concept.

**Comment by CompCodeNet** [comment]
@jwakely I can see that inline wrappers for contained objects would essentially be "comments" since the compiler would turn them into NOPs except the actual _containedObject.SomeFunction( ) call. I can also see that it's a decent approach to keeping code clean and simple. (The only other point that I was making is that it's good id the language can accommodate these things in syntax and thus reduce the verbosity.) I also encourage people to be more welcoming of ideas, especially good ones that have value and something that all C++ developers would be grateful to see, or at least explain why it's not a good idea. I'm not a rookie, I started using "OO" in the 80's using function pointers on structs before C++ was even available and I started using C++ as soon as it was available.
